pkg/querier/tripperware/metadata/split_by_interval.go

```diff
@@ -0,0 +1,81 @@
+package metadata
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/cortexproject/cortex/pkg/querier/tripperware"
+)
+
+type IntervalFn func(r tripperware.Request) time.Duration
+
+// SplitByIntervalMiddleware creates a new Middleware that splits requests by a given interval.
+func SplitByIntervalMiddleware(interval IntervalFn, limits tripperware.Limits, merger tripperware.Merger, registerer prometheus.Registerer) tripperware.Middleware {
```

Comment:
Those are also duplicate code as the split middleware in query range. What about reusing the same one instead of maintaining another middleware?

Reply:
These middlewares work quite differently:
- The step param is not used for splitting metadata APIs.
- The requests are merged as they come in for metadata APIs.

This new middleware will be shared for all metadata APIs.

Reply:
But we can still reuse the same one, right? Just cast the request and see if it is query range request or not. If not, we don't need to handle steps. That's what Thanos does.